src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/DefaultLSPProgressListener.cs

```diff
@@ -0,0 +1,171 @@
+            {
+                return;
+            }
+
+            var token = parameterToken["token"].ToObject<string>(); // IProgress<object>>();
+
+            if (string.IsNullOrEmpty(token) || !_activeRequests.TryGetValue(token, out var request))
+            {
+                return;
+            }
+
+            var value = parameterToken["value"];
+
+            try
+            {
+                await request.OnProgressResult(value).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Initial handle request has been cancelled
+                // We can deem this ProgressRequest complete
+                ProgressCompleted(token);
+                return;
+            }
+
+            var cts = new CancellationTokenSource();
+
+            lock (request.RequestLock)
+            {
+                var existingCTS = request.CancellationTokenSource;
+                if (existingCTS != null &&
+                    existingCTS.Token.CanBeCanceled &&
+                    !existingCTS.Token.IsCancellationRequested)
+                {
+                    existingCTS.Cancel();
+                }
+
+                request.CancellationTokenSource = cts;
+            }
+
+            _ = CompleteAfterDelayAsync(token, request); // Fire and forget
+        }
+
+        private async Task CompleteAfterDelayAsync(string token, ProgressRequest request)
+        {
+            if (request.CancellationTokenSource.IsCancellationRequested)
+            {
+                return;
+            }
+
+            try
+            {
+                await Task.Delay(request.TimeoutAfterLastNotify, request.CancellationTokenSource.Token).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Task cancelled, new progress notification received.
+                // Don't allow handler to return
+                return;
+            }
+
+            ProgressCompleted(token);
+        }
+
+        internal override bool TryListenForProgress(
+            string requestId,
+            Func<JToken, Task> onProgressResult,
+            TimeSpan timeoutAfterLastNotify,
+            out Task onCompleted)
+        {
+            var onCompletedSource = new TaskCompletionSource<bool>();
+            var progressRequest = new ProgressRequest(onProgressResult, timeoutAfterLastNotify, onCompletedSource);
+            onCompleted = onCompletedSource.Task;
+
+            return _activeRequests.TryAdd(requestId, progressRequest);
+        }
+
+        private void ProgressCompleted(string requestId)
+        {
+            if (_activeRequests.TryRemove(requestId, out var request))
+            {
+                // We're setting the result of a Task<bool>
+                // however we only return a Task to the
+                // handler/subscriber, so the bool is ignored
+                request.OnCompleted.SetResult(false);
+            }
```

Comment:
Should we add an `else` Debug.Fail clause? I wouldn't imagine that'd ever be expected right?

Reply:
I see the else case happening:
1. Two progress notifications arrive at the same time to the `ClientNotifyAsync`
2. They are both processed (lets say sequentially) https://github.com/dotnet/aspnetcore-tooling/pull/2119/files/0fa823f2b6218c9ea96d02c4028b3b20a521104a#diff-ef97568d908d1a85a8be9a7a91c660c0R69
3. Let's say the original task has been cancelled, so both should throw the Task Cancelled Exception and result in us calling `ProgressCompleted`. 
4. The first will remove the request from `_activeRequests`. The second `TryRemove` will fail.

In this scenario, IMO the best option would just be to allow it to pass through without a `Debug.Fail`. This is a situation I do see happening, and there's no harm (as the "unregistration" has already happened once).

Reply:
Ahh ya totally agree then, no debug.fail!